demo/queued_agent_win.cc

```diff
@@ -0,0 +1,270 @@
+// Copyright 2022 The Chromium Authors.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include <windows.h>
+
+#include <fstream>
+#include <iostream>
+#include <queue>
+#include <string>
+
+#include "content_analysis/sdk/analysis_agent.h"
+
+using content_analysis::sdk::Agent;
+using content_analysis::sdk::Session;
+using content_analysis::sdk::ContentAnalysisRequest;
+using content_analysis::sdk::ContentAnalysisResponse;
+
+// This class maintains a list of outstanding content analysis requests to
+// process.  Each request is encapsulated in one Session.  Requests are handled
+// in FIFO order.
+class RequestQueue {
+ public:
+  RequestQueue() {
+    Init();
+  }
+  ~RequestQueue() {
+    Term();
+  }
+
+  // Push a new content analysis session into the queue.
+  void push(std::unique_ptr<Session> session) {
+    Enter();
+    sessions_.push(std::move(session));
+    WakeOne();
+    Leave();
+  }
+
+  // Pop the next request from the queue, blocking if necessary until a session
+  // is available.  Returns a nullptr if the queue will produce no more
+  // sessions.
+  std::unique_ptr<Session> pop() {
+    Enter();
+
+    while (!abort_ && sessions_.size() == 0)
+      Wait();
+
+    std::unique_ptr<Session> session;
+    if (!abort_) {
+      session = std::move(sessions_.front());
+      sessions_.pop();
+    }
+
+    Leave();
+    return session;
+  }
+
+  // Marks the queue as aborted.  pop() will now return nullptr.
+  void abort() {
+    Enter();
+    abort_ = true;
+    WakeAll();
+    Leave();
+  }
+
+ private:
+  void Init() {
+    InitializeConditionVariable(&cv_);
+    InitializeCriticalSection(&cs_);
+  }
+  void Term() {
+  }
+  void Enter() {
+    EnterCriticalSection(&cs_);
+  }
+  void Leave() {
+    LeaveCriticalSection(&cs_);
+  }
+  void Wait() {
+    SleepConditionVariableCS(&cv_, &cs_, INFINITE);
+  }
+  void WakeOne() {
+    WakeConditionVariable(&cv_);
+  }
+  void WakeAll() {
+    WakeAllConditionVariable(&cv_);
+  }
+
+  CRITICAL_SECTION cs_;
```

Comment:
nit: i'd put {} before the semicolon so that these members are value-initialized.

Reply:
Done.